### PR TITLE
Adding api for cm data flow request

### PR DESCRIPTION
**Pull request by Nexengineer**

### CODE REVIEW

File: src/In.ProjectEKA.HipService/DataFlow/Database/Migrations/20200204100834_AddingDataFlowRequestTable.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
+﻿using Microsoft.EntityFrameworkCore.Migrations;
+
+namespace In.ProjectEKA.HipService.DataFlow.Database.Migrations
+{
+    public partial class AddingDataFlowRequestTable : Migration
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/DataFlow/Database/Migrations/20200204100834_AddingDataFlowRequestTable.cs]: Codacy Issue found: ['partial' is gratuitous in this context.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4941971)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- src/In.ProjectEKA.HipService/DataFlow/HealthInformationResponse.cs  1
- src/In.ProjectEKA.HipService/DataFlow/Model/DataFlowRequest.cs  1
- src/In.ProjectEKA.HipService/DataFlow/Database/Migrations/DataFlowContextModelSnapshot.cs  1
- test/In.ProjectEKA.HipServiceTest/DataFlow/DataFlowControllerTest.cs  1
- src/In.ProjectEKA.HipService/DataFlow/DataFlow.cs  2
- test/In.ProjectEKA.HipServiceTest/DataFlow/Builder/TestBuilder.cs  1
- test/In.ProjectEKA.HipServiceTest/DataFlow/DataFlowRepositoryTest.cs  1
- src/In.ProjectEKA.HipService/DataFlow/HiDataRange.cs  1
- src/In.ProjectEKA.HipService/DataFlow/Consent.cs  1
- src/In.ProjectEKA.HipService/DataFlow/Database/DataFlowContext.cs  1
- src/In.ProjectEKA.HipService/DataFlow/Database/Migrations/20200204100834_AddingDataFlowRequestTable.cs  1
- src/In.ProjectEKA.HipService/DataFlow/Database/Migrations/20200204100834_AddingDataFlowRequestTable.Designer.cs  1
- src/In.ProjectEKA.HipService/DataFlow/DataFlowRepository.cs  1
- src/In.ProjectEKA.HipService/DataFlow/HealthInformationRequest.cs  1
- test/In.ProjectEKA.HipServiceTest/DataFlow/DataFlowTest.cs  1
- src/In.ProjectEKA.HipService/DataFlow/DataFlowController.cs  2
         

Clones added
============
- src/In.ProjectEKA.HipService/DataFlow/Database/Migrations/DataFlowContextModelSnapshot.cs  1
- src/In.ProjectEKA.HipService/DataFlow/Database/Migrations/20200204100834_AddingDataFlowRequestTable.Designer.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4941971&bid=16337639)

File: src/In.ProjectEKA.HipService/DataFlow/DataFlow.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+            return result.Map(r =>
+            {
+                var errorResponse = new ErrorResponse(new Error(ErrorCode.ServerInternalError,
+                    ErrorMessage.InternalServerError));
+                return new Tuple<HealthInformationResponse, ErrorResponse>(null, errorResponse);
+            }).ValueOr(new Tuple<HealthInformationResponse,
+                ErrorResponse>(new HealthInformationResponse(request.TransactionId), null));
+        }
+
+        private static bool ValidateConsentArtefact(Consent consent)
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/DataFlow/DataFlow.cs]: Codacy Issue found: [Remove this unused method parameter 'consent'.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4941971)

File: src/In.ProjectEKA.HipService/DataFlow/DataFlow.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        private readonly IDataFlowRepository dataFlowRepository;
+
+        public DataFlow(IDataFlowRepository dataFlowRepository)
+        {
+            this.dataFlowRepository = dataFlowRepository;
+        }
+        
+        public async Task<Tuple<HealthInformationResponse, ErrorResponse>> HealthInformationRequestFor(HealthInformationRequest request)
+        {
+            if (!ValidateConsentArtefact(request.Consent))
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/DataFlow/DataFlow.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4941971)

File: test/In.ProjectEKA.HipServiceTest/DataFlow/Builder/TestBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
+using In.ProjectEKA.HipService.DataFlow;
+
+namespace In.ProjectEKA.HipServiceTest.DataFlow.Builder
+{
+    using Bogus;
+
+    public class TestBuilder
```
Comment (ganesanarun) [inline test/In.ProjectEKA.HipServiceTest/DataFlow/Builder/TestBuilder.cs]: Codacy Issue found: [Add a 'protected' constructor or the 'static' keyword to the class declaration.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4941971)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- test/In.ProjectEKA.HipServiceTest/DataFlow/DataFlowRepositoryTest.cs  1
- src/In.ProjectEKA.HipService/DataFlow/Consent.cs  1
- src/In.ProjectEKA.HipService/DataFlow/HiDataRange.cs  1
- test/In.ProjectEKA.HipServiceTest/DataFlow/Builder/TestBuilder.cs  1
- test/In.ProjectEKA.HipServiceTest/DataFlow/DataFlowControllerTest.cs  1
- src/In.ProjectEKA.HipService/DataFlow/DataFlowRepository.cs  1
- src/In.ProjectEKA.HipService/DataFlow/Database/Migrations/20200204100834_AddingDataFlowRequestTable.Designer.cs  1
- src/In.ProjectEKA.HipService/DataFlow/DataFlowController.cs  2
- src/In.ProjectEKA.HipService/DataFlow/Model/DataFlowRequest.cs  1
- src/In.ProjectEKA.HipService/DataFlow/DataFlow.cs  2
- test/In.ProjectEKA.HipServiceTest/DataFlow/DataFlowTest.cs  1
- src/In.ProjectEKA.HipService/DataFlow/HealthInformationResponse.cs  1
- src/In.ProjectEKA.HipService/DataFlow/Database/DataFlowContext.cs  1
- src/In.ProjectEKA.HipService/DataFlow/Database/Migrations/DataFlowContextModelSnapshot.cs  1
- src/In.ProjectEKA.HipService/DataFlow/HealthInformationRequest.cs  1
- src/In.ProjectEKA.HipService/DataFlow/Database/Migrations/20200204100834_AddingDataFlowRequestTable.cs  1
         

Clones added
============
- src/In.ProjectEKA.HipService/DataFlow/Database/Migrations/20200204100834_AddingDataFlowRequestTable.Designer.cs  1
- src/In.ProjectEKA.HipService/DataFlow/Database/Migrations/DataFlowContextModelSnapshot.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4941971&bid=16337639)

File: src/In.ProjectEKA.HipService/DataFlow/Database/Migrations/20200204100834_AddingDataFlowRequestTable.cs (hunk 1 above)
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/DataFlow/Database/Migrations/20200204100834_AddingDataFlowRequestTable.cs]: Codacy Issue found: ['partial' is gratuitous in this context.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4941971)

File: src/In.ProjectEKA.HipService/DataFlow/DataFlow.cs [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+            return result.Map(r =>
+            {
+                var errorResponse = new ErrorResponse(new Error(ErrorCode.ServerInternalError,
+                    ErrorMessage.InternalServerError));
+                return new Tuple<HealthInformationResponse, ErrorResponse>(null, errorResponse);
+            }).ValueOr(new Tuple<HealthInformationResponse,
+                ErrorResponse>(new HealthInformationResponse(request.TransactionId), null));
+        }
+
+        private static bool ValidateConsentArtefact(Consent consent)
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/DataFlow/DataFlow.cs]: Codacy Issue found: [Remove this unused method parameter 'consent'.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4941971)

File: src/In.ProjectEKA.HipService/DataFlow/DataFlow.cs [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        private readonly IDataFlowRepository dataFlowRepository;
+
+        public DataFlow(IDataFlowRepository dataFlowRepository)
+        {
+            this.dataFlowRepository = dataFlowRepository;
+        }
+        
+        public async Task<Tuple<HealthInformationResponse, ErrorResponse>> HealthInformationRequestFor(HealthInformationRequest request)
+        {
+            if (!ValidateConsentArtefact(request.Consent))
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/DataFlow/DataFlow.cs]:
```suggestion
            if (!IsValidateConsentArtefact(request.Consent))
```

File: src/In.ProjectEKA.HipService/DataFlow/DataFlow.cs [hunk 7]
```diff
@@ -0,0 +1,39 @@
+namespace In.ProjectEKA.HipService.DataFlow
+{
+    using System;
+    using System.Threading.Tasks;
+    using HipLibrary.Patient.Model.Response;
+    
+    public class DataFlow: IDataFlow
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/DataFlow/DataFlow.cs]:
```suggestion
    public class DataFlow : IDataFlow
```

File: src/In.ProjectEKA.HipService/DataFlow/DataFlow.cs [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+        public DataFlow(IDataFlowRepository dataFlowRepository)
+        {
+            this.dataFlowRepository = dataFlowRepository;
+        }
+        
+        public async Task<Tuple<HealthInformationResponse, ErrorResponse>> HealthInformationRequestFor(HealthInformationRequest request)
+        {
+            if (!ValidateConsentArtefact(request.Consent))
+            {
+                return new Tuple<HealthInformationResponse, ErrorResponse>(null, null);
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/DataFlow/DataFlow.cs]: Why error is also null?

File: src/In.ProjectEKA.HipService/DataFlow/DataFlow.cs [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+            }
+            
+            var result = await dataFlowRepository.SaveRequestFor(request.TransactionId, request)
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/DataFlow/DataFlow.cs]: If the transactionId is not unique, will there be an exception? If so, where are we handling it?

File: src/In.ProjectEKA.HipService/DataFlow/DataFlowRepository.cs [hunk 10]
```diff
@@ -0,0 +1,33 @@
+namespace In.ProjectEKA.HipService.DataFlow
+{
+    using System;
+    using System.Threading.Tasks;
+    using Database;
+    using Model;
+    using Optional;
+    public class DataFlowRepository: IDataFlowRepository
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/DataFlow/DataFlowRepository.cs]:
```suggestion
    public class DataFlowRepository : IDataFlowRepository
```

File: src/In.ProjectEKA.HipService/DataFlow/DataFlowRepository.cs [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    {
+        private readonly DataFlowContext dataFlowContext;
+
+        public DataFlowRepository(DataFlowContext dataFlowContext)
+        {
+            this.dataFlowContext = dataFlowContext;
+        }
+
+        public async Task<Option<Exception>> SaveRequestFor(string transactionId,
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/DataFlow/DataFlowRepository.cs]:
```suggestion
        public async Task<Option<Exception>> SaveRequestFor(string transactionId,
```
```suggestion
        public async Task<Option<Exception>> SaveRequestFor(
              string transactionId,
```

Comment (ganesanarun) [reply src/In.ProjectEKA.HipService/DataFlow/DataFlowRepository.cs]: Format properly.

File: src/In.ProjectEKA.HipService/DataFlow/HiDataRange.cs [hunk 12]
```diff
@@ -0,0 +1,14 @@
+namespace In.ProjectEKA.HipService.DataFlow
+{
+    public class HiDataRange
+    {
+        public string From { get; }
+        public string To { get; }
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/DataFlow/HiDataRange.cs]: Why can't it be datetime?

File: src/In.ProjectEKA.HipService/DataFlow/KeyMaterial.cs [hunk 13]
```diff
@@ -0,0 +1,7 @@
+namespace In.ProjectEKA.HipService.DataFlow
+{
+    public class KeyMaterial
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/DataFlow/KeyMaterial.cs]: Remove this for now. We can figure out this as part of encryption.

File: src/In.ProjectEKA.HipService/DataFlow/Model/DataFlowRequest.cs [hunk 14]
```diff
@@ -0,0 +1,21 @@
+namespace In.ProjectEKA.HipService.DataFlow.Model
+{
+    using System.ComponentModel.DataAnnotations;
+
+    public class DataFlowRequest
+    {
+        [Key] public string TransactionId { get; }
+        
+        public HealthInformationRequest HealthInformationRequest { get; }
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/DataFlow/Model/DataFlowRequest.cs]: HealthInformationRequest itself has a transactionID, why one more transactionID?

File: src/In.ProjectEKA.HipService/In.ProjectEKA.HipService.csproj [hunk 15]
```diff
@@ -29,6 +29,7 @@
     </ItemGroup>
 
     <ItemGroup>
+      <Folder Include="DataFlow\Database\Migrations" />
       <Folder Include="Discovery\Database\Migrations" />
     </ItemGroup>
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/In.ProjectEKA.HipService.csproj]: Is it required?

File: test/In.ProjectEKA.HipServiceTest/DataFlow/Builder/TestBuilder.cs [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+{
+    using Bogus;
+
+    public static class TestBuilder
+    {
+        private static Faker faker;
+        
+        internal static Faker Faker() => faker ??= new Faker();
+        
+        internal static HealthInformationRequest HealthInformationRequest(string transactionId)
```
Comment (ganesanarun) [inline test/In.ProjectEKA.HipServiceTest/DataFlow/Builder/TestBuilder.cs]: Can't we use Faker<HealthInformationRequest>?

Comment (Nexengineer) [reply test/In.ProjectEKA.HipServiceTest/DataFlow/Builder/TestBuilder.cs]: tried using it but it comes as null

File: src/In.ProjectEKA.HipService/DataFlow/Model/DataFlowRequest.cs (hunk 14 above)
Comment (Nexengineer) [reply src/In.ProjectEKA.HipService/DataFlow/Model/DataFlowRequest.cs]: It is acting as a key

File: src/In.ProjectEKA.HipService/DataFlow/HiDataRange.cs (hunk 12 above)
Comment (Nexengineer) [reply src/In.ProjectEKA.HipService/DataFlow/HiDataRange.cs]: It is stored as a string in consent manager

File: src/In.ProjectEKA.HipService/DataFlow/DataFlow.cs (hunk 6 above)
Comment (Nexengineer) [reply src/In.ProjectEKA.HipService/DataFlow/DataFlow.cs]: done

File: src/In.ProjectEKA.HipService/DataFlow/DataFlow.cs (hunk 9 above)
Comment (Nexengineer) [reply src/In.ProjectEKA.HipService/DataFlow/DataFlow.cs]: Yeah it is handled in repositry

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- src/In.ProjectEKA.HipService/DataFlow/HiDataRange.cs  1
- src/In.ProjectEKA.HipService/DataFlow/Model/DataFlowRequest.cs  1
- src/In.ProjectEKA.HipService/DataFlow/Database/Migrations/20200204100834_AddingDataFlowRequestTable.cs  1
- test/In.ProjectEKA.HipServiceTest/DataFlow/Builder/TestBuilder.cs  1
- src/In.ProjectEKA.HipService/DataFlow/HealthInformationRequest.cs  1
- test/In.ProjectEKA.HipServiceTest/DataFlow/DataFlowRepositoryTest.cs  1
- src/In.ProjectEKA.HipService/DataFlow/DataFlowRepository.cs  1
- src/In.ProjectEKA.HipService/DataFlow/Database/Migrations/DataFlowContextModelSnapshot.cs  1
- src/In.ProjectEKA.HipService/DataFlow/Database/Migrations/20200204100834_AddingDataFlowRequestTable.Designer.cs  1
- test/In.ProjectEKA.HipServiceTest/DataFlow/DataFlowControllerTest.cs  1
- src/In.ProjectEKA.HipService/DataFlow/DataFlow.cs  2
- src/In.ProjectEKA.HipService/DataFlow/DataFlowController.cs  2
- src/In.ProjectEKA.HipService/DataFlow/HealthInformationResponse.cs  1
- src/In.ProjectEKA.HipService/DataFlow/Consent.cs  1
- src/In.ProjectEKA.HipService/DataFlow/Database/DataFlowContext.cs  1
- test/In.ProjectEKA.HipServiceTest/DataFlow/DataFlowTest.cs  1
         

Clones added
============
- src/In.ProjectEKA.HipService/DataFlow/Database/Migrations/DataFlowContextModelSnapshot.cs  1
- src/In.ProjectEKA.HipService/DataFlow/Database/Migrations/20200204100834_AddingDataFlowRequestTable.Designer.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4941971&bid=16337639)

File: src/In.ProjectEKA.HipService/DataFlow/Database/Migrations/20200204100834_AddingDataFlowRequestTable.cs (hunk 1 above)
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/DataFlow/Database/Migrations/20200204100834_AddingDataFlowRequestTable.cs]: Codacy Issue found: ['partial' is gratuitous in this context.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4941971)

File: src/In.ProjectEKA.HipService/DataFlow/DataFlow.cs [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+            return result.Map(r =>
+            {
+                var errorResponse = new ErrorResponse(new Error(ErrorCode.ServerInternalError,
+                    ErrorMessage.InternalServerError));
+                return new Tuple<HealthInformationResponse, ErrorResponse>(null, errorResponse);
+            }).ValueOr(new Tuple<HealthInformationResponse,
+                ErrorResponse>(new HealthInformationResponse(request.TransactionId), null));
+        }
+
+        private static bool IsValidConsentArtefact(Consent consent)
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/DataFlow/DataFlow.cs]: Codacy Issue found: [Remove this unused method parameter 'consent'.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=4941971)

**Pull request merged**